### PR TITLE
docs: improve page send and spend funds from predicates

### PR DESCRIPTION
**Pull request by Torres-ssf**
This PR aims to:
- Rewrite the doc page [send and spend funds from predicates](https://fuellabs.github.io/fuels-ts/guide/predicates/send-and-spend-funds-from-predicates.html)
- Add tested code snippets to this doc page.
- Remove old code snippets related to the same doc page.

### CODE REVIEW

**Review by danielbate — DISMISSED** [review]

**Comment by danielbate** [comment]
LGTM 👍🏻 Approving as the following suggesting may be out of scope of this PR. I've seen a couple mentions in the form for _cancelling_ a predicate. 

- https://forum.fuel.network/t/how-to-send-money-from-predicate-to-owner-back/2639
- https://forum.fuel.network/t/how-to-get-asset-outputs-for-amount-and-get-asset-inputs-for-amount-with-ts-dsk/2638

Should we extend this page to include that case?

**Comment by Torres-ssf** [comment]
> LGTM 👍🏻 Approving as the following suggesting may be out of scope of this PR. I've seen a couple mentions in the form for _cancelling_ a predicate.
> 
> * https://forum.fuel.network/t/how-to-send-money-from-predicate-to-owner-back/2639
> * https://forum.fuel.network/t/how-to-get-asset-outputs-for-amount-and-get-asset-inputs-for-amount-with-ts-dsk/2638
> 
> Should we extend this page to include that case?

@danielbate I think these two topics might not be directly related to this page's content.

The first one seems to be an unexpected error and I guess it's not related to the configurable constants. I'm planning to try and replicate it locally to understand better.

The second one is about canceling a created predicate. I wasn't aware that this could be done and I'm not sure how we'd do it in the TS SDK. It feels like this topic deserves its own page.

**Comment by danielbate** [comment]
@Torres-ssf Sorry yes, although the posts themselves aren't specifically about cancelling predicates they do make reference to it. I'm also unsure about the recommended process for doing so, if someone could inform me of the process (if it's even possible) I'd be happy to raise an issue for it.

**Comment by arboleya** [comment]
@danielbate What does it mean to "cancel a predicate"?

@Torres-ssf Few observations:
 1. Why did you duplicate this file inside the `docs-snippets` instead of referencing the [original](https://github.com/FuelLabs/fuels-ts/blob/master/packages/fuel-gauge/test-projects/predicate-triple-sig/src/main.sw) from `fuel-gauge`? The same goes for the related [unit test](https://github.com/FuelLabs/fuels-ts/blob/master/packages/fuel-gauge/src/doc-examples.test.ts#L324). In short, I don't think we should be duplicating stuff inside the `docs-snippets`.
 3. This Sway example seems over-complex; could we remove the triple-sig stuff to make it easier to understand?

**Comment by danielbate** [comment]
> @danielbate What does it mean to "cancel a predicate"?

I will reach out to Composability Labs for their exact use case but I interpreted it as the sender wallet being able to retrieve the funds from the predicate that were originally intended for a receiver via a predicate. Granted the sender wallet could evaluate the predicate also, potentially that is the recommended process?

**Comment by Torres-ssf** [comment]
> @danielbate What does it mean to "cancel a predicate"?
> 
> @Torres-ssf Few observations:
> 
> 1. Why did you duplicate this file inside the `docs-snippets` instead of referencing the [original](https://github.com/FuelLabs/fuels-ts/blob/master/packages/fuel-gauge/test-projects/predicate-triple-sig/src/main.sw) from `fuel-gauge`? The same goes for the related [unit test](https://github.com/FuelLabs/fuels-ts/blob/master/packages/fuel-gauge/src/doc-examples.test.ts#L324). In short, I don't think we should be duplicating stuff inside the `docs-snippets`.
> 2. This Sway example seems over-complex; could we remove the triple-sig stuff to make it easier to understand?

@arboleya The predicates are not equal. The original predicate seems to be extracting the data input not from the params of the `main` function but from a function imported from the standard lib: ` inputs::input_predicate_data`. I tried to reproduce this example locally, but could not verify if this is something possible. Not sure if I am missing something or if this approach does not work anymore. 

The test related to this predicate on the `fuel gauge` uses a different JSON abi, a hardcoded one. I could verify that the generated JSON abi from the original predicate in the fuel-gauge, results in a different JSON abi from the one that is hardcoded and being used on the test file where the previous code snippets were placed.

And yes now that you mentioned it, I do think that this example is too complex. I think it makes sense to give a simpler example.

**Comment by danielbate** [comment]
> @danielbate Wouldn't this be a withdrawal?
> 
> We have this unit test:
> 
> https://github.com/FuelLabs/fuels-ts/blob/04a4b3122ecb502747b50f5df64c5e61b39e80cc/packages/fuel-gauge/src/doc-examples.test.ts#L441

But `withdraw` is a method defined in the contract:

https://github.com/FuelLabs/fuels-ts/blob/8c42ba22035404aa49ea36546e4c896bd65c36f7/packages/fuel-gauge/test-projects/liquidity-pool/src/main.sw#L23-L24

Rather than functionality baked into a predicate. So by that definition, are we saying that _cancellation_ isn't a possible state. And that the withdrawal can only be made by the sender evaluating the predicate in the same way that any receiver would. And if a consumer wants to provide withdrawals, this should be written into the predicate evaluation logic.

Apologies also, still getting up to scratch with predicates.

**Comment by arboleya** [comment]
@danielbate You're right, apologies. I didn't notice that the unit test I sent interacted with a `Contract` instead of a `Predicate` and that the `withdraw` was a user-defined method. I guess I was just confused by the "cancel" terminology.

I'm not well-versed in Predicates, but your reasoning makes sense.

File: apps/docs-snippets/projects/simple-predicate/src/main.sw [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// #region send-and-spend-funds-from-predicates-1
+predicate;
+
+fn main(inputted_address: b256) -> bool {
+    let valid_address = 0xfc05c23a8f7f66222377170ddcbfea9c543dff0dd2d2ba4d0478a4521423a9d4;
+
+    inputted_address == valid_address
```
[the remark this answers is not in the document]
Comment (arboleya) [reply apps/docs-snippets/projects/simple-predicate/src/main.sw:7]: Nice catch, `received_address` would be another option. Pick your flavor.

**Review by danielbate — APPROVED** [review]

**Pull request merged**